### PR TITLE
[RAC-6339] Support iPDU's SFTP Server  

### PR DESCRIPTION
**Pull request by HQebupt**
To support the iPDU firmware update of ServerTech, fix sftp-job.js to upload the firmware image.

- Relate PR :  [on-taskgraph-312](https://github.com/RackHD/on-taskgraph/pull/312)

### CODE REVIEW

File: lib/jobs/sftp-job.js [hunk 1]
```diff
@@ -25,11 +26,13 @@ function sftpJobFactory(
     Promise,
     cryptService,
     ssh,
+    fs,
     waterline
 ) {
     var logger = Logger.initialize(sftpJobFactory);
 
     function SftpJob(options, context, taskId) {
+        console.log("===sftp-job.js, options:", options);
```
Comment (panpan0000) [inline lib/jobs/sftp-job.js]: remove debug code.

File: package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "//": "Copyright 2017, Dell EMC, Inc.",
   "name": "on-tasks",
-  "version": "2.28.0",
+  "version": "2.29.0",
```
Comment (panpan0000) [inline package.json]: please rebase again to remove this change.

File: lib/jobs/sftp-job.js (hunk 1 above)
Comment (HQebupt) [reply lib/jobs/sftp-job.js]: 3x, I will remove all the debug code.

Comment (pengz1) [inline lib/jobs/sftp-job.js]: Remove this line

File: lib/jobs/sftp-job.js [hunk 3]
```diff
@@ -44,13 +47,16 @@ function sftpJobFactory(
         this.keepaliveCountMax = options.keepaliveCountMax;
         this.fileSource = options.fileSource;
         this.fileDestination = options.fileDestination;
+        this.isPDU= options.isPDU;
     }
     util.inherits(SftpJob, BaseJob);
 
     SftpJob.prototype._run = function() {
+        console.log('======= SftpJob.run:=======');
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: Remove this line.

File: lib/jobs/sftp-job.js [hunk 4]
```diff
@@ -44,13 +47,16 @@ function sftpJobFactory(
...
         var self = this;
         return waterline.ibms.findByNode(this.nodeId, 'ssh-ibm-service')
         .then(function(sshSettings) {
+            console.log("======sshSettings:", sshSettings);
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: Remove this line.

File: lib/jobs/sftp-job.js [hunk 5]
```diff
@@ -67,19 +73,36 @@ function sftpJobFactory(
 
             var ssh = sshClient;
             ssh.on('ready', function() {
+                console.log('====== SftpJob.ssh: ready');
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: Remove this line.

File: lib/jobs/sftp-job.js [hunk 6]
```diff
@@ -67,19 +73,36 @@ function sftpJobFactory(
...
                 ssh.sftp(function(err, sftp) {
                     if (err) { reject(err); }
-                    sftp.fastPut(
+                    if (self.isPDU === "true") {
+                        console.log("==== this is an iPDU, isPDU:", self.isPDU);
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: Remove this line.

File: lib/jobs/sftp-job.js [hunk 7]
```diff
@@ -67,19 +73,36 @@ function sftpJobFactory(
...
+                        var readStream = fs.createReadStream(self.fileSource);
+                        var writeStream = sftp.createWriteStream(self.fileDestination);
+                        writeStream.on('close',function () {
+                            console.log( "===== file transferred succesfully" );
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: You should use logger.debug if you want to record something. And also, you should use a callback to handle error case. And in the callback function, reject and resolve should be called.

File: lib/jobs/sftp-job.js [hunk 8]
```diff
@@ -67,19 +73,36 @@ function sftpJobFactory(
...
+                        });
+                        writeStream.on('end', function () {
+                            console.log( "===== sftp connection closed" );
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: You should use logger.debug if you want to record something. And also, you should use a callback to handle error case. And in the callback function, reject and resolve should be called.

File: lib/jobs/sftp-job.js [hunk 9]
```diff
@@ -67,19 +73,36 @@ function sftpJobFactory(
...
+                            ssh.close();
+                        });
+                        readStream.pipe( writeStream );
+                    } else {
+                        console.log("==== normal sftp server, isPDU:", self.isPDU);
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: Remove this line.

File: lib/jobs/sftp-job.js [hunk 10]
```diff
@@ -67,19 +73,36 @@ function sftpJobFactory(
...
                                 if(err) {  reject(err); }
                                 ssh.end();
                             }
-                    );
+                        );
+                    }
                 });
             })
             .on('error', function(err) {
+                console.log('====== SftpJob.ssh: error');
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: Remove this line.

File: lib/jobs/sftp-job.js [hunk 11]
```diff
@@ -89,6 +112,7 @@ function sftpJobFactory(
                 reject(err);
             })
             .on('close', function(hasErr) {
+                console.log('====== SftpJob.ssh: close');
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: Remove this line.

File: lib/jobs/sftp-job.js [hunk 12]
```diff
@@ -110,6 +134,17 @@ function sftpJobFactory(
                 port: sshSettings.port || 22,
                 username: sshSettings.user,
                 password: cryptService.decrypt(sshSettings.password),
+
+                debug: function(info){
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: Remote those lines that are for debugging.

File: lib/task-data/tasks/sftp-task.js [hunk 13]
```diff
@@ -1,12 +1,13 @@
 // Copyright 2016, EMC, Inc.
 
 'use strict';
-
+console.log('=====call sftp-task.js');
```
Comment (pengz1) [inline lib/task-data/tasks/sftp-task.js]: Remove this line.

File: package.json (hunk 2 above)
Comment (pengz1) [inline package.json]: It is not necessary to change this line.

**Comment by pengz1** [comment]
Since you have a new if-else branch, unit test is required to test that branch.

File: lib/jobs/sftp-job.js [hunk 14]
```diff
@@ -69,14 +73,54 @@ function sftpJobFactory(
...
+                                    destination: self.fileDestination,
+                                    host: sshSettings.host
+                                });
+
+                                reject(new Error(
+                                    "Encountered a failure transferring "+self.fileSource+
+                                    "to "+self.fileDestination+"on remote host"+ sshSettings.host
+                                ));
+                            } else {
+                                resolve();
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: You can have more than one reject, but there should only be on resolve.

File: lib/jobs/sftp-job.js [hunk 15]
```diff
@@ -69,14 +73,54 @@ function sftpJobFactory(
             ssh.on('ready', function() {
                 ssh.sftp(function(err, sftp) {
                     if (err) { reject(err); }
-                    sftp.fastPut(
+                    if (self.isPDU === "true") {
+                        logger.info("This is an iPDU, isPDU:", self.isPDU);
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: There are two much logger messages, actually this is not necessary. Usually we only log important millstones inside a task.

File: lib/jobs/sftp-job.js [hunk 16]
```diff
@@ -69,14 +73,54 @@ function sftpJobFactory(
...
+                                    destination: self.fileDestination,
+                                    host: sshSettings.host
+                                });
+
+                                reject(new Error(
+                                    "Encountered a failure transferring end "+self.fileSource+
+                                    "to "+self.fileDestination+"on remote host"+ sshSettings.host
+                                ));
+                            } else {
+                                resolve();
```
Comment (pengz1) [inline lib/jobs/sftp-job.js]: resolve() should be the last line of callback function.

**Comment by panpan0000** [comment]
:+0

@anhou  @pengz1 , can you please help to review ASAP?

File: spec/lib/jobs/sftp-job-spec.js [hunk 17]
```diff
@@ -199,6 +209,55 @@ describe('sftp-job', function() {
...
+                    },
+                    privateKey: 'a pretty long, encrypted string',
+                    keepaliveInterval: 3000,
+                    keepaliveCountMax: 4,
+                    readyTimeout: 30000
+                });
+            });
+        });
+
+        it.only('should pass isPDU options to iPDU firmware update', function() {
```
Comment (anhou) [inline spec/lib/jobs/sftp-job-spec.js]: remove '.only'

Comment (pengz1) [inline spec/lib/jobs/sftp-job-spec.js]: Remove only

**Pull request merged**